src/journal/bsod.c

```diff
@@ -0,0 +1,143 @@
+        if (r < 0) {
+                log_warning_errno(r, "Failed to open journal: %m");
+                return NULL;
+        }
+
+        r = sd_id128_get_boot(&boot_id);
+        if (r < 0) {
+                log_warning_errno(r, "Failed to get boot ID, ignoring : %m");
+                goto clean_journal;
+        }
+        sd_id128_to_string(boot_id, boot_id_string);
+
+        snprintf(boot_id_filter, sizeof(boot_id_filter), "_BOOT_ID=%s", boot_id_string);
+        r = sd_journal_add_match(j, boot_id_filter, 0);
+        if (r < 0) {
+                log_warning_errno(r, "Failed to add boot ID filter: %m");
+                goto clean_journal;
+        }
+
+        r = sd_journal_add_match(j, "_UID=0", 0);
+        if (r < 0) {
+                log_warning_errno(r, "Failed to add User ID filter: %m");
+                goto clean_journal;
+        }
+
+        r = sd_journal_add_match(j, "PRIORITY=0", 0);
+        if (r < 0) {
+                log_warning_errno(r, "Failed to add Emergency filter: %m");
+                goto clean_journal;
+        }
+
+        SD_JOURNAL_FOREACH(j) {
+                r = sd_journal_get_data(j, "MESSAGE", &d, &l);
+                if (r < 0) {
+                        log_warning_errno(r, "Failed to read journal message: %m");
+                        continue;
+                }
+                message = strndup((char *) d, l);
+                break;
+        }
+
+clean_journal:
+        sd_journal_close(j);
+        return message;
+}
+
+static int parse_argv(int argc, char * argv[]) {
+
+        enum {
+                ARG_VERSION = 0x100,
+        };
+
+        static const struct option options[] = {
+                { "help",    no_argument, NULL, 'h'         },
+                { "version", no_argument, NULL, ARG_VERSION },
+                {}
+        };
+
+        int c;
+
+        assert(argc >= 0);
+        assert(argv);
+
+        while ((c = getopt_long(argc, argv, "h", options, NULL)) >= 0)
+
+                switch (c) {
+
+                case 'h':
+                        help();
+                        return 0;
+
+                case ARG_VERSION:
+                        return version();
+
+                case '?':
+                        return -EINVAL;
+
+                default:
+                        assert_not_reached();
+                }
+        if (optind < argc)
+                return log_error_errno(SYNTHETIC_ERRNO(EINVAL),
+                                       "%s takes no argument.",
+                                       program_invocation_short_name);
+        return 1;
+}
+
+static int run(int argc, char *argv[]) {
+        int r;
+
+        log_open();
+        log_parse_environment();
+
+        r = parse_argv(argc, argv);
+        if (r <= 0)
+                return r;
+
+        _cleanup_close_ int fd = -EBADF;
+        char * message = first_emerg_boot_message();
+
+        fd = open_terminal("/dev/console", O_WRONLY|O_NOCTTY|O_CLOEXEC);
```

Comment:
so for this I think we should not go to /dev/console (which is a magic device that basically points to wherever boot time logs and stuff should go to which can be serial or similar) but the VT subsystem (which is the virtual terminal, i.e. a textual interface to a local physical display device, i.e. no serial), and I think we should not interfere with the usual log/progress output that /dev/console gets. Hence I'd suggest we allocate a full new VT for this, and then switch to that.

(in case you wonder what a VT is, it's this archaic textual display logic that the linux kernel uses to do early boot logging before wayland/x11 take over, and that you can log into via Alt-F2, Alt-F3, …)

i.e. open /dev/tty1 temporarily, then issue the VT_GETSTATE ioctl call on it which tells you which VTs are currently allocated via a bitmask. Look for the first free VT (i.e. determine lowest unset bit, then format /dev/tty%i with it plus one. Then open that, and use that. switch to it via the VT_ACTIVATE ioctl, then clear it by output ANSI_HOME_CLEAR on it, then display the message there.

If this sounds like a bit much, grep our sources for VT_GETSTAT, we already call that ioctl elsewhere for other reasons.

Reply:
`VT_OPENQRY` might be easier than `VT_GETSTATE` and `__builtin_ctz(~vt_stat.v_state)` or whatever.